### PR TITLE
Fix preprocessor handling with sibling errors

### LINKED ISSUE #2687: Preprocessing hides errors in zod 3.22.x
Consider this piece of code:

```typescript
import { z } from "zod";

const zodStringWithPreprocess =
  // Notice that in this simplified example, the preprocessing does nothing
  z.preprocess((val) => val, z.string()) as z.ZodEffects<
    z.ZodString,
    string,
    string
  >;

const validatorWithoutPreprocess = z.object({
  name: z.string(),
  town: z.string(),
});

const validatorWithPreprocess = z.object({
  name: zodStringWithPreprocess,
  town: zodStringWithPreprocess,
});

const dataToValidate = {};

const parsedWithoutPreprocess =
  validatorWithoutPreprocess.safeParse(dataToValidate);
console.log("Without preprocess");
console.log(JSON.stringify(parsedWithoutPreprocess, null, 2));

const parsedWithPreprocess = validatorWithPreprocess.safeParse(dataToValidate);
console.log("With preprocess");
console.log(JSON.stringify(parsedWithPreprocess, null, 2));
```

# Expected behavior

In Zod 3.21.4, the result is the same regardless of preprocessing:

```
Without preprocess
{
  "success": false,
  "error": {
    "issues": [
      {
        "code": "invalid_type",
        "expected": "string",
        "received": "undefined",
        "path": [
          "name"
        ],
        "message": "Required"
      },
      {
        "code": "invalid_type",
        "expected": "string",
        "received": "undefined",
        "path": [
          "town"
        ],
        "message": "Required"
      }
    ],
    "name": "ZodError"
  }
}
With preprocess
{
  "success": false,
  "error": {
    "issues": [
      {
        "code": "invalid_type",
        "expected": "string",
        "received": "undefined",
        "path": [
          "name"
        ],
        "message": "Required"
      },
      {
        "code": "invalid_type",
        "expected": "string",
        "received": "undefined",
        "path": [
          "town"
        ],
        "message": "Required"
      }
    ],
    "name": "ZodError"
  }
}
```

# Actual behavior

In Zod 3.22.x (x <= 2), the version with preprocessing stops at the first error:

```text
Without preprocess
{
  "success": false,
  "error": {
    "issues": [
      {
        "code": "invalid_type",
        "expected": "string",
        "received": "undefined",
        "path": [
          "name"
        ],
        "message": "Required"
      },
      {
        "code": "invalid_type",
        "expected": "string",
        "received": "undefined",
        "path": [
          "town"
        ],
        "message": "Required"
      }
    ],
    "name": "ZodError"
  }
}
With preprocess
{
  "success": false,
  "error": {
    "issues": [
      {
        "code": "invalid_type",
        "expected": "string",
        "received": "undefined",
        "path": [
          "name"
        ],
        "message": "Required"
      }
    ],
    "name": "ZodError"
  }
}
```

Notice that the `"expected": "string",` for the `town` property is missing.

It seems to be a regression for Zod.

### PR DESCRIPTION
**Pull request by kpdecker**
Fixes #2687

### CODE REVIEW

File: deno/lib/README.md [hunk 1]
```diff
@@ -721,7 +721,7 @@ z.string().regex(regex);
 z.string().includes(string);
 z.string().startsWith(string);
 z.string().endsWith(string);
-z.string().datetime(); // defaults to UTC, see below for options
+z.string().datetime(); // ISO 8601; default is without UTC offset, see below for options
```
Comment (kpdecker) [inline deno/lib/README.md:724]: Unsure if PRs should include deno build artifacts. If no I can pull.

**Comment by thiagomeireless** [comment]
Hey @colinhacks would you be able to merge this fix? It's affecting multiple projects...

**Comment by azhiv** [comment]
I confirm that this PR also fixes [#2677](https://github.com/colinhacks/zod/issues/2677) and [#2671](https://github.com/colinhacks/zod/issues/2671).

**Comment by rap2hpoutre** [comment]
I would be glad having it merged! DO you have a fork @kpdecker

**Comment by moroine** [comment]
Hey, any update on this?

Do we plan to merge this? I can help if needed.

**Comment by flavianh** [comment]
Please merge this, I can't upgrade Zod otherwise 😅

**Comment by flavianh** [comment]
@colinhacks Is there something blocking this PR? Can we help somehow?

**Comment by micahh2** [comment]
It seems like (according to this comment: https://github.com/colinhacks/zod/issues/2820#issuecomment-1747627127), the team here uses "thumbs up on the root comment" as a measure to gauge popularity of a PR.

If you'd like to see this PR considered, it could help to give a :+1: to the first comment on this thread.

**Comment by aaronadamsCA** [comment]
@kpdecker As a non-contributor who would love to see this merged, may I suggest some improvements to reduce the change size?

#2912 is a duplicate by @yukukotani with a simpler fix:

https://github.com/colinhacks/zod/pull/2912/files#diff-c54113cf61ec99691748a3890bfbeb00e10efb3f0a76f03a0fd9ec49072e410a

#3124 is another duplicate by @selimb with a simpler test (which itself could be further simplified):

https://github.com/colinhacks/zod/pull/3124/files#diff-baf84a3000a84c7d96342171542949faad6d2d5d28830f610af2df10180e3b9f

I'd also suggest removing the unrelated `README.md` changes.

**Comment by aaronadamsCA** [comment]
Here is a one-line fix you can apply as a local patch to the Zod package:

- [For Deno users](https://github.com/colinhacks/zod/pull/2912/files#diff-52632a4861fc9d7dc2dacef13cd91d60286dd706c1bb57438b8ee6a579a8796a)
- [For everyone else](https://github.com/colinhacks/zod/pull/2912/files#diff-c54113cf61ec99691748a3890bfbeb00e10efb3f0a76f03a0fd9ec49072e410a)

I would suggest lending your 👍 to #2912 to encourage a fix to be merged.

**Comment by nurbolnygmanov** [comment]
> I unfortunately don't have time to support this PR as we've transitioned away from Zod. Closing to let one of the active community members take over.

what did you choose as an alternative?

**Pull request closed without merging**